controller/scheduler-v2/job.go

```diff
@@ -0,0 +1,179 @@
+package main
+
+import (
+	"errors"
+	"time"
+
+	ct "github.com/flynn/flynn/controller/types"
+	"github.com/flynn/flynn/controller/utils"
+	"github.com/flynn/flynn/host/types"
+)
+
+type JobRequestType string
+type JobState string
+
+const (
+	JobRequestTypeUp   JobRequestType = "up"
+	JobRequestTypeDown JobRequestType = "down"
+	JobStateStarting                  = "starting"
+	JobStateRunning                   = "running"
+	JobStateStopped                   = "stopped"
+	JobStateCrashed                   = "crashed"
+	JobStateRequesting                = "requesting"
+	JobStateNew                       = "new"
+)
+
+type JobRequest struct {
+	*Job
+	RequestType JobRequestType
+	attempts    uint
+}
+
+func NewJobRequest(f *Formation, requestType JobRequestType, typ, hostID, jobID string) *JobRequest {
+	return &JobRequest{
+		Job:         NewJob(f, f.App.ID, f.Release.ID, typ, hostID, jobID),
+		RequestType: requestType,
+	}
+}
+
+func (r *JobRequest) needsVolume() bool {
+	return r.Job.Formation.Release.Processes[r.Type].Data
+}
+
+func (r *JobRequest) Clone() *JobRequest {
+	return &JobRequest{
+		Job:         r.Job.Clone(),
+		RequestType: r.RequestType,
+		attempts:    r.attempts,
+	}
+}
+
+type Job struct {
+	Type      string
+	AppID     string
+	ReleaseID string
+	HostID    string
+	JobID     string
+
+	Formation *Formation
+
+	restarts  uint
+	startedAt time.Time
+	state     JobState
+}
+
+func NewJob(f *Formation, appID, releaseID, typ, hostID, id string) *Job {
+	return &Job{
+		Type:      typ,
+		AppID:     appID,
+		ReleaseID: releaseID,
+		HostID:    hostID,
+		JobID:     id,
+		Formation: f,
+		startedAt: time.Now(),
+		state:     JobStateNew,
+	}
+}
+
+func (j *Job) Clone() *Job {
+	// Shallow copy
+	cloned := *j
+	return &cloned
+}
+
+func (j *Job) IsScheduled() bool {
+	return j.state != JobStateStopped
+}
+
+func (j *Job) IsRunning() bool {
+	return j.state == JobStateStarting || j.state == JobStateRunning
+}
+
+func (j *Job) IsSchedulable() bool {
+	return j.Formation != nil && j.Type != ""
+}
+
+type Jobs map[string]*Job
+
+func (js Jobs) GetStoppableJobs(key utils.FormationKey, typ string) []*Job {
+	formTypeJobs := make([]*Job, 0, len(js))
+	for _, j := range js {
+		if j.IsScheduled() && j.Formation != nil && j.Formation.key() == key && j.Type == typ {
+			formTypeJobs = append(formTypeJobs, j)
+		}
+	}
+	return formTypeJobs
+}
+
+func (js Jobs) GetHostJobCounts(key utils.FormationKey, typ string) map[string]int {
+	counts := make(map[string]int)
+
+	for _, j := range js {
+		if j.IsScheduled() && j.Formation != nil && j.Formation.key() == key && j.Type == typ {
+			counts[j.HostID]++
+		}
+	}
+	return counts
+}
+
+func (js Jobs) GetProcesses(key utils.FormationKey) Processes {
+	procs := make(Processes)
+	for _, j := range js {
+		if j.IsScheduled() && j.Formation != nil && j.Formation.key() == key {
+			procs[j.Type]++
+		}
+	}
+	return procs
+}
+
+func (js Jobs) AddJob(j *Job) *Job {
+	j = j.Clone()
```

Comment:
I did this to remove some bugs with job IDs getting changed in a job that was in `s.jobs`. I think copy-on-write is a good idea anyway, since mutating something referenced elsewhere is risky. I can probably remove the clone since I think the path where it was needed no longer exists.